front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,897 @@
+        },
+        {} as Record<string, string[]>
+      );
+
+    for (const trig of Object.keys(trigramAggreg)) {
+      if (trigramAggreg[trig].length < 2) {
+        delete trigramAggreg[trig];
+      }
+      trigramAggreg[trig] = sortBy(trigramAggreg[trig], (key) => {
+        if (key.startsWith('op_id:')) return 1;
+        if (key.startsWith('trigram:')) return 2;
+        if (key.startsWith('uic:')) return 3;
+        // default
+        return 4;
+      });
+    }
+
+    Object.values(trigramAggreg).forEach((mergeList) => {
+      const mainNodeKey = mergeList[0];
+      mergeList.slice(1).forEach((key) => {
+        this.nodesByPathKey[key] = mainNodeKey;
+      });
+    });
+  }
+
+  /**
+   * Store and index the node.
+   */
+  private indexNode(node: MacroNodeResponse, saved = false) {
+    // Remove in the id index, its previous value
+    const currentValue = this.getNodeByKey(node.path_item_key);
+    if (currentValue && typeof currentValue !== 'string') {
+      const prevId = currentValue.node.id;
+      delete this.nodesIdToKey[prevId];
+    }
+
+    // Index
+    this.nodesByPathKey[node.path_item_key] = { node, saved };
+    this.nodesIdToKey[node.id] = node.path_item_key;
+    node.labels.forEach((l) => {
+      if (l) this.labels.add(l);
+    });
+  }
+
+  /**
+   * Update node's data by it's key
+   */
+  private updateNodeData(key: string, data: Partial<Omit<MacroNodeResponse, 'path_item_key'>>) {
+    const indexedNode = this.getNodeByKey(key);
+    if (indexedNode) {
+      this.indexNode({ ...indexedNode.node, ...data }, indexedNode.saved);
+    }
+  }
+
+  /**
+   * Delete a node by it's key
+   */
+  private deleteNode(key: string) {
+    const indexedNode = this.getNodeByKey(key);
+    if (indexedNode) {
+      delete this.nodesIdToKey[indexedNode.node.id];
+      delete this.nodesByPathKey[key];
+    }
+  }
+
+  /**
+   * Get a node by its id.
+   */
+  private getNodeById(id: number) {
+    const key = this.nodesIdToKey[id];
+    return this.getNodeByKey(key);
+  }
+
+  /**
+   * Get a node by its key.
+   */
+  private getNodeByKey(key: string): NodeIndex | null {
+    let result: NodeIndex | null = null;
+    let currentKey: string | null = key;
+    while (currentKey !== null) {
+      const found: string | NodeIndex = this.nodesByPathKey[currentKey];
+      if (typeof found === 'string') {
+        currentKey = found;
+      } else {
+        currentKey = null;
+        result = found;
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Apply a layout on nodes and save the new position.
+   * Nodes that are saved are fixed.
+   */
+  private applyLayout() {
+    const graph = this.toGraph();
+    if (graph.nodes.some((n) => !n.fixed)) {
+      const layout = new cola.Layout()
+        .defaultNodeSize(100)
+        .size([800, 600])
+        .linkDistance(200)
+        .avoidOverlaps(true)
+        .nodes(graph.nodes)
+        .links(graph.edges)
+        .start(10, 15, 20, 0, false, false);
```

Comment:
This is not what we've decided in https://github.com/osrd-project/osrd-confidential/issues/658

Reply:
I had many issue on node's position, specially when I add a new nodes (ex: adding a complete new train schedule), or when nodes were closed each other (sometimes a node was completely hidden by an other one).

By checking the ticket you mention, and by talking with @louisgreiner I implemented a simple version of the web cola algo.

Mixing geospatial coordinates with cartesian, when a user can add new nodes & move nodes and their positions are locked,  is tricky. It asks more question than it solves issues. 
But adding a geospatial view of the chart is easy.

Exemple: a user move all the nodes and don't respect the geo ( from what I understood, it is the standard use-case). Then he adds a new train station.
Where should it be added  ? At its translated geoposition ?  Ok, but then what is the geo center of the chart in that case ?

Reply:
That's all been discussed in the linked issue already.

This PR is big enough as-is, discussing nodes position just increases the size further and begs more questions: is the implemented behavior really what we want? Does it work fine in all edge cases?

I'd recommend leaving this for later. If you really want to keep this in, put in a separate commit.